### DISCUSSION TITLE
Stop using tsconfig paths to locally link packages in the workspace

### DISCUSSION THREAD

**Discussion by ajwootto** [nrwl/nx#23339] [Feature Requests]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

Wrote in more detail in this discussion:
https://github.com/nrwl/nx/discussions/20005

I wanted to start a discussion on whether Nx could ever stop using its `paths` linking strategy and switch to native workspace package linking.

## Motivation
Basically Nx uses the "paths" in tsconfig to link packages in the workspace together. It then magically rewrites those paths to corresponding `dist/` directories when things are built. That system has a bunch of problems described in the discussion like relying on "relative" modules (a legacy feature), as well as requiring plugins for a bunch of tools that could otherwise work natively with workspace linking built into package managers. It also causes a lot of very painful problems when the magic goes wrong, that are very difficult to debug without diving around Nx's source and adding logs to figure out what's happening.
There are also other downsides, like the fact that "nx serve" ignores the tsconfig of the dependent packages being linked, whereas "nx build" usually respects them:
https://github.com/nrwl/nx/issues/10614

## Suggested Implementation

No idea, seems really hard :)

Some resources I found about different ways to do this:
https://turbo.build/blog/you-might-not-need-typescript-project-references
https://thijs-koerselman.medium.com/my-quest-for-the-perfect-ts-monorepo-62653d3047eb

**Comment by FrozenPandaz** [comment]
I've converted this to a discussion so it can be discussed more easily.

**Comment by ajwootto** [comment]
This is also the cause of this issue, which I don't really see a way of working around without changing how Nx does things
https://github.com/nrwl/nx/issues/18324

Due to the reliance on relative module imports, you basically _can't_ use the Node16 moduleResolution setting with ESM in Nx
